unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java

```diff
@@ -106,92 +168,125 @@ public int length() {
    * @param until the position after last code point, exclusive.
    */
   public UTF8String substring(final int start, final int until) {
-    if (until <= start || start >= bytes.length) {
+    if (until <= start || start >= numBytes) {
       return UTF8String.fromBytes(new byte[0]);
     }
 
     int i = 0;
     int c = 0;
-    for (; i < bytes.length && c < start; i += numBytes(bytes[i])) {
+    for (; i < numBytes && c < start; i += numBytesForFirstByte(getByte(i))) {
       c += 1;
     }
 
     int j = i;
-    for (; j < bytes.length && c < until; j += numBytes(bytes[i])) {
+    for (; j < numBytes && c < until; j += numBytesForFirstByte(getByte(i))) {
       c += 1;
     }
 
-    return UTF8String.fromBytes(Arrays.copyOfRange(bytes, i, j));
+    byte[] bytes = new byte[j - i];
+    copyMemory(base, offset + i, bytes, BYTE_ARRAY_OFFSET, j - i);
+    return UTF8String.fromBytes(bytes);
   }
 
   public boolean contains(final UTF8String substring) {
-    final byte[] b = substring.getBytes();
-    if (b.length == 0) {
+    if (substring.numBytes == 0) {
       return true;
     }
 
-    for (int i = 0; i <= bytes.length - b.length; i++) {
-      if (bytes[i] == b[0] && startsWith(b, i)) {
+    byte first = substring.getByte(0);
+    for (int i = 0; i <= numBytes - substring.numBytes; i++) {
+      if (getByte(i) == first && matchAt(substring, i)) {
         return true;
       }
     }
     return false;
   }
 
-  private boolean startsWith(final byte[] prefix, int offsetInBytes) {
-    if (prefix.length + offsetInBytes > bytes.length || offsetInBytes < 0) {
+  private byte getByte(int i) {
+    return UNSAFE.getByte(base, offset + i);
+  }
+
+  private boolean matchAt(final UTF8String s, int pos) {
+    if (s.numBytes + pos > numBytes || pos < 0) {
       return false;
     }
-    int i = 0;
-    while (i < prefix.length && prefix[i] == bytes[i + offsetInBytes]) {
-      i++;
-    }
-    return i == prefix.length;
+    return ByteArrayMethods.arrayEquals(base, offset + pos, s.base, s.offset, s.numBytes);
   }
 
   public boolean startsWith(final UTF8String prefix) {
-    return startsWith(prefix.getBytes(), 0);
+    return matchAt(prefix, 0);
   }
 
   public boolean endsWith(final UTF8String suffix) {
-    return startsWith(suffix.getBytes(), bytes.length - suffix.getBytes().length);
+    return matchAt(suffix, numBytes - suffix.numBytes);
   }
 
   public UTF8String toUpperCase() {
-    return UTF8String.fromString(toString().toUpperCase());
+    // It's always have the same number of bytes for upper case
+    byte[] buf = new byte[numBytes];
+    copyMemory(base, offset, buf, BYTE_ARRAY_OFFSET, numBytes);
+
+    for (int i = 0; i < numBytes; ){
+      int n = numBytesForFirstByte(getByte(i));
+      int code = codePointAt(i, n);
+      int upper = Character.toUpperCase(code);
```

Comment:
Using ScalaCheck, I think I found some failing test cases.  I'll open a PR against this to demonstrate in a sec.